pallets/inflation-rewards/src/lib.rs

```diff
@@ -0,0 +1,316 @@
+    }
+
+    /// Pending authors per block and per container chain
+    #[pallet::storage]
+    #[pallet::getter(fn pending_rewards)]
+    pub(super) type PendingAuthors<T: Config> = StorageDoubleMap<
+        _,
+        Blake2_128Concat,
+        BlockNumber,
+        Twox64Concat,
+        ParaId,
+        PendingAuthorsValue<T>,
+        OptionQuery,
+    >;
+
+    /// Information extracted from a container chain header
+    #[derive(
+        Clone, Encode, Decode, PartialEq, sp_core::RuntimeDebug, scale_info::TypeInfo, MaxEncodedLen,
+    )]
+    #[scale_info(skip_type_params(T))]
+    pub struct PendingAuthorsValue<T: Config> {
+        pub authors: BoundedVec<(T::AccountId, u32), T::MaxAuthors>,
+        pub last_block_number: BlockNumber,
+    }
+    impl<T: Config> PendingAuthorsValue<T> {
+        pub(super) fn new(author: T::AccountId, block_number: BlockNumber) -> Self {
+            let mut authors = BoundedVec::new();
+            authors
+                .try_push((author, 1))
+                .expect("MaxAuthors should be greather than zero");
+            Self {
+                authors,
+                last_block_number: block_number,
+            }
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        fn reward_orchestrator_author() -> Weight {
+            let mut total_weight = T::DbWeight::get().reads(1);
+            let orchestrator_author = T::GetSelfChainBlockAuthor::get();
+
+            if let Some(chains_to_reward) = ChainsToReward::<T>::get() {
+                total_weight += T::DbWeight::get().reads(1);
+                match T::StakingRewardsDistributor::distribute_rewards(
+                    orchestrator_author,
+                    T::Currency::withdraw(
+                        &T::PendingRewardsAccount::get(),
+                        chains_to_reward.rewards_per_chain,
+                        Precision::BestEffort,
+                        Preservation::Expendable,
+                        Fortitude::Force,
+                    )
+                    .unwrap_or(CreditOf::<T>::zero()),
+                ) {
+                    Ok(frame_support::dispatch::PostDispatchInfo {
+                        actual_weight: Some(weight),
+                        ..
+                    }) => total_weight += weight,
+                    Err(e) => {
+                        log::debug!("Fail to distribute rewards: {:?}", e)
+                    }
+                    _ => {}
+                }
+            } else {
+                panic!("ChainsToReward not filled");
+            }
+
+            total_weight
+        }
+    }
+}
+
+impl<T: Config> AuthorNotingHook<T::AccountId> for Pallet<T> {
+    fn on_container_author_noted(
+        author: &T::AccountId,
+        block_number: BlockNumber,
+        para_id: ParaId,
+    ) -> Weight {
+        let mut total_weight = T::DbWeight::get().reads_writes(1, 1);
+        let pending_authors_per_chain = if let Some(mut pending_authors_per_chain) =
```

Comment:
how would we have something to read here if we are writing this value later on this function? I am not sure I understand this function, so comments would be good

Reply:
We cannot have two authors for the same block number for the same para id right?